### ISSUE TITLE
[RFC] Decouple indices from database specifics

### ISSUE THREAD

**Issue by gctucker** [kernelci/kernelci-api#437]
The way indices are currently created is by calling a class method in the model, for example:
```python
    @classmethod
    def create_indexes(cls, collection):
        collection.create_index("name", unique=True)
```

However, this means the model expects the `collection` to be a PyMongo object so it ties the model to the database engine.

While looking at how to implement something similar for another use-case, I came up with this solution which I think could be used here as well:
```python
class Model(BaseModel):
    @dataclass
    class Index:
        collection: str
        field: str|list
        attributes: dict[str, Any]
```

Then in the database implementation:
```python
class Database:
    models: [models.Thing]

    def __init__(self, mongo):
        self._client = pymongo.MongoClient(mongo)
        self._create_indices()

    def _create_indices(self):
        for model in self.models:
            for index in model.get_indices():
                col = self._client[index.collection]
                col.create_index(index.field, **index.attributes)
```

In this case, the attributes are passed as-is so it kind of expects the models to "know" about MongoDB attributes.  Still this could be translated if needed.  So here's an example of how indices would be defined:
```python
class Thing(Model):

    name: str

    @classmethod
    def get_indices(cls) -> list[Model.Index]:
        return [
            Model.Index('things', 'name', {'unique': True}),
        ]
```

The collection name could also be removed from Index and stored in the Model instead, so models don't define indexes for other collections.  But that's more of an implementation detail, the key thing here is that the model doesn't need to import `pymongo`.  I think this might also help with keeping models in `kernelci-core` although indices aren't exposed via the API of course, it's more something for internal API models.  It can definitely help if adding support for multiple database backends e.g. `sqlite` and I think it's good practice in general to decouple the models.

Hope this helps!

**Comment by gctucker** [comment]
Note: to remove the collection from Index:
```python
class Database:
    models = {models.Thing: 'things'}
    def _create_indices(self):
        for collection, model in self.models.items():
            col = self._client[collection]
            col.create_index(index.field, **index.attributes)
```
which I think is closer to the current `Database.COLLECTIONS` dictionary.

**Comment by JenySadadia** [comment]
Thanks for the proposal @gctucker .
I have started looking into the implementation.

**Comment by hardboprobot** [comment]
It's good to decouple the design from the underlying DB but this is very low priority. After all, the whole application is still tightly coupled to mongodb, if we wanted to abstract the DB access away there's a lot more work to do than this, and are we planning to switch DBMS any time soon?

**Comment by gctucker** [comment]
I agree it's very low priority, probably something to do later in 2024 once the initial version has been well stabilised in production.

There is a potential use-case for standalone deployments where MongoDB is overkill.  But also I went through a similar FastAPI use-case in a different project and thought it would be worth considering here as it's still a minor code rework at this stage.  It could be done as a low-priority thing in parallel to high-priority ones I guess, clearing up technical debt is always good and there's still lots of it from the early PoC implementation.

**Comment by gctucker** [comment]
Also it's worth mentioning that indices should not be defined as part of the user-facing models.  So the models in `kernelci-core` should not have this information, it's only something specific to the API internal implementation.

**Comment by JenySadadia** [comment]
As I have already done some work in this direction, can we consider reviewing https://github.com/kernelci/kernelci-api/pull/446, and then we can drop it if it doesn't seem required atm?
Bdw, I also agree that decoupling should be done at an early stage.
